lib/validators/url_validator.rb

```diff
@@ -26,21 +29,28 @@ def initialize(options)
           else options[:scheme]
           end
 
-        options.reverse_merge!(message: I18n.t('errors.messages.invalid_url'))
+        options[:message] ||= I18n.t('errors.messages.invalid_url')
         super(options)
       end
 
       def validate_each(record, attribute, value)
-        url = URI(value.to_s)
         valid = catch(:invalid) do
-          validate_domain(value)
-          validate_scheme(options[:scheme], url.scheme) if options.key?(:scheme)
-          validate_path(options[:path], url.path) if options.key?(:path)
-          validate_query(options[:query], url.query) if options.key?(:query)
-          validate_fragment(options[:fragment], url.fragment) if options.key?(:fragment)
-          true # valid
+            url = URI(value.to_s)
+            if accept_relative_urls?
+              validate_domain_absense(url)
+            else
+              validate_domain(value)
+              validate_authority(options[:authority], url) if options.key?(:authority)
+              validate_scheme(options[:scheme], url.scheme) if options.key?(:scheme)
+            end
+            validate_path(options[:path], url.path) if options.key?(:path)
+            validate_query(options[:query], url.query) if options.key?(:query)
+            validate_fragment(options[:fragment], url.fragment) if options.key?(:fragment)
+            true # valid
+          # end
         end
-
+      rescue URI::InvalidURIError
```

Comment:
Instead of catch-blocks, perhaps we can throw `URI::InvalidURIError` in the `validate_*` methods instead?  In that case, we don't need the `ensure` line, nor the `catch` construct, nor the `unless valid` check.